grin/src/Grin/TypeCheck.hs

```diff
@@ -31,62 +34,69 @@ import Data.Bifunctor
          - fixed node tag arity
     done - monomorph nodes ; each node component is
             - singleton simple type
-            - one or more location
+            - one or more locations
+            - unspecified location
+            - dead type (empty simple typeset)
 -}
 typeEnvFromHPTResult :: HPTResult -> Either String TypeEnv.TypeEnv
 typeEnvFromHPTResult hptResult = typeEnv where
-  convertSimpleType :: SimpleType -> TypeEnv.SimpleType
+  convertSimpleType :: SimpleType -> Either String TypeEnv.SimpleType
   convertSimpleType = \case
-    T_Int64   -> TypeEnv.T_Int64
-    T_Word64  -> TypeEnv.T_Word64
-    T_Float   -> TypeEnv.T_Float
-    T_Bool    -> TypeEnv.T_Bool
-    T_Unit    -> TypeEnv.T_Unit
-    T_Location l -> TypeEnv.T_Location [l]
+    T_Int64               -> pure TypeEnv.T_Int64
+    T_Word64              -> pure TypeEnv.T_Word64
+    T_Float               -> pure TypeEnv.T_Float
+    T_Bool                -> pure TypeEnv.T_Bool
+    T_Unit                -> pure TypeEnv.T_Unit
+    T_UnspecifiedLocation -> pure TypeEnv.T_UnspecifiedLocation
+    T_Location l          -> pure $ TypeEnv.T_Location [l]
+    l@Local{} -> throwError $ "Encountered analysis specific local value during type checking: " ++ show l
```

Comment:
Couldn't we decouple analysis specific values from results? I feel that the final HPTResult should not contain abstractions like "Local"

Reply:
Local abstract values are analysis specific artifacts. They can be used to give easy-to-understand names to these artifacts, thus differentiating them from the generic values. Also, local values should not escape the scope of the given analysis. They should be interpreted by the analysis, and transformed away.